### PR TITLE
NSX Integration fixes

### PR DESCRIPTION
**Pull request by Pearl1594**
### Description

This PR addresses issues identified in https://github.com/apache/cloudstack/pull/7919 

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test matrix

**Comment by Pearl1594** [comment]
@blueorangutan package

**Review by nvazquez — APPROVED** [review]

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan package

File: plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterService.java [hunk 1]
```diff
@@ -92,6 +92,30 @@ public interface KubernetesClusterService extends PluggableService, Configurable
             true,
             ConfigKey.Scope.Account,
             KubernetesServiceEnabled.key());
+    static final ConfigKey<Long> KubernetesControlNodeInstallAttemptWait = new ConfigKey<Long>("Advanced", Long.class,
```
Comment (nvazquez) [inline plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterService.java:95]: I think these settings shouldn't be part of this PR but for #9102

File: plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/actionworkers/KubernetesClusterStartWorker.java [hunk 2]
```diff
@@ -241,6 +248,11 @@ private String getKubernetesAdditionalControlNodeConfig(final String joinIp, fin
         final String sshPubKey = "{{ k8s.ssh.pub.key }}";
         final String clusterHACertificateKey = "{{ k8s_control_node.cluster.ha.certificate.key }}";
         final String ejectIsoKey = "{{ k8s.eject.iso }}";
+        final String installWaitTime = "{{ k8s.install.wait.time }}";
```
Comment (nvazquez) [inline plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/actionworkers/KubernetesClusterStartWorker.java:251]: Same for all these CKS logic using the settings values

File: plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterService.java (hunk 1 above)
Comment (Pearl1594) [reply plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterService.java:95]: I believe this is just externalizing the already present in the CKS scripts. So it wouldn't cause any issues. Unless you still want to me to revert

Comment (nvazquez) [reply plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterService.java:95]: IMO we should fix only NSX related issues on this PR and we could add the externalize settings on the other PR

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test matrix

**Comment by Pearl1594** [comment]
@nvazquez this is good to be reviewed and merged - I have added the fix for the test failures wrt host affinity - not sure why it was failing on this PR - its unrelated to the changes made. All the failures seen on this PR are again unrelated and seem to be environmental or due to cleanup issues.

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test matrix

File: engine/components-api/src/main/java/com/cloud/network/vpc/VpcManager.java [hunk 3]
```diff
@@ -122,7 +122,7 @@ public interface VpcManager {
      * @throws InsufficientAddressCapacityException
      * @throws ConcurrentOperationException
      */
-    PublicIp assignSourceNatIpAddressToVpc(Account owner, Vpc vpc) throws InsufficientAddressCapacityException, ConcurrentOperationException;
+    PublicIp assignSourceNatIpAddressToVpc(Account owner, Vpc vpc, Long podId, boolean forNsx) throws InsufficientAddressCapacityException, ConcurrentOperationException;
```
Comment (DaanHoogland) [inline engine/components-api/src/main/java/com/cloud/network/vpc/VpcManager.java:125]: how about creating a default parameter and calling an overloaded method.
In addition to that `forNsx` seems very implementation specific; can it be retrieved somehow from configuration and hidden from the service interface?

File: plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/actionworkers/KubernetesClusterResourceModifierActionWorker.java [hunk 4]
```diff
@@ -542,10 +552,11 @@ protected void removePortForwardingRules(final IpAddress publicIp, final Network
         for (PortForwardingRuleVO pfRule : pfRules) {
             if (startPort <= pfRule.getSourcePortStart() && pfRule.getSourcePortStart() <= endPort) {
                 portForwardingRulesDao.remove(pfRule.getId());
-                logger.debug("The Port forwarding rule [{}] with the id [{}] was removed.", pfRule.getName(), pfRule.getId());
+                logger.trace("Marking PF rule " + pfRule + " with Revoke state");
```
Comment (DaanHoogland) [inline plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/actionworkers/KubernetesClusterResourceModifierActionWorker.java:555]: this log message is a bit of a regression, does it really need to change?

File: plugins/network-elements/nsx/src/main/java/org/apache/cloudstack/resource/NsxResource.java [hunk 5]
```diff
@@ -385,16 +393,21 @@ private NsxAnswer executeRequest(CreateNsxPortForwardRuleCommand cmd) {
                 cmd.getNetworkResourceId(), cmd.isResourceVpc());
         try {
             String privatePort = cmd.getPrivatePort();
-            String service = privatePort.contains("-") ? nsxApiClient.getServicePath(ruleName, privatePort, cmd.getProtocol(), null, null) :
-                    nsxApiClient.getNsxInfraServices(ruleName, privatePort, cmd.getProtocol(), null, null);
+            logger.debug(String.format("Checking if rule %s exists on Tier 1 Gateway: %s", ruleName, tier1GatewayName));
```
Comment (DaanHoogland) [inline plugins/network-elements/nsx/src/main/java/org/apache/cloudstack/resource/NsxResource.java:396]:
```suggestion
            logger.debug("Checking if rule {} exists on Tier 1 Gateway: {}", ruleName, tier1GatewayName);
```

File: plugins/network-elements/nsx/src/main/java/org/apache/cloudstack/service/NsxApiClient.java [hunk 6]
```diff
@@ -199,6 +205,26 @@ public NsxApiClient(String hostname, String port, String username, char[] passwo
...
+                return false;
+            }
+            ControllerClusterStatus status = clusterStatus.getControlClusterStatus();
+            if (status == null) {
+                logger.error("Cannot get NSX Controller Cluster Status");
+                return false;
+            }
+            return CLUSTER_STATUS_STABLE.equalsIgnoreCase(status.getStatus());
+        } catch (Error error) {
+            logger.error(String.format("Error checking NSX Controller Health: %s", error.getMessage()));
```
Comment (DaanHoogland) [inline plugins/network-elements/nsx/src/main/java/org/apache/cloudstack/service/NsxApiClient.java:223]:
```suggestion
            logger.error("Error checking NSX Controller Health: {}", error.getMessage());
```

File: plugins/network-elements/nsx/src/main/java/org/apache/cloudstack/service/NsxApiClient.java [hunk 7]
```diff
@@ -485,13 +513,16 @@ private PolicyGroupMembersListResult getSegmentPortList(SegmentPorts segmentPort
...
+        int retries = NsxService.NSX_API_FAILURE_RETRIES.valueIn(zoneId);
+        int waitingSecs = NsxService.NSX_API_FAILURE_INTERVAL.valueIn(zoneId);
         int count = 1;
+        Long portCount;
         do {
             try {
-                logger.info("Waiting for all port groups to be unlinked from the segment - Attempt: " + count++ + " Waiting for 5 secs");
-                Thread.sleep(5000);
+                logger.info(String.format("Waiting for all port groups to be unlinked from the segment %s - " +
+                        "Attempt: %s. Waiting for %s secs", segmentName, count++, waitingSecs));
```
Comment (DaanHoogland) [inline plugins/network-elements/nsx/src/main/java/org/apache/cloudstack/service/NsxApiClient.java:524]:
```suggestion
                logger.info("Waiting for all port groups to be unlinked from the segment {} - " +
                        "Attempt: {}. Waiting for {} secs", segmentName, count++, waitingSecs);
```

File: plugins/network-elements/nsx/src/main/java/org/apache/cloudstack/service/NsxApiClient.java [hunk 8]
```diff
@@ -526,24 +557,37 @@ public void createStaticNatRule(String vpcName, String tier1GatewayName,
...
             NatRules natService = (NatRules) nsxService.apply(NatRules.class);
-            logger.debug(String.format("Deleting NSX static NAT rule %s for tier-1 gateway %s (network: %s)", ruleName, tier1GatewayName, networkName));
-            // delete NAT rule
-            natService.delete(tier1GatewayName, NatId.USER.name(), ruleName);
-            if (service == Network.Service.PortForwarding) {
-                String svcName = getServiceName(ruleName, privatePort, protocol, null, null);
-                // Delete service
-                Services services = (Services) nsxService.apply(Services.class);
-                services.delete(svcName);
+            logger.debug(String.format("Deleting NSX NAT rule %s for tier-1 gateway %s (network: %s)", ruleName, tier1GatewayName, networkName));
```
Comment (DaanHoogland) [inline plugins/network-elements/nsx/src/main/java/org/apache/cloudstack/service/NsxApiClient.java:578]:
```suggestion
            logger.debug("Deleting NSX NAT rule {} for tier-1 gateway {} (network: {})", ruleName, tier1GatewayName, networkName);
```

File: plugins/network-elements/nsx/src/main/java/org/apache/cloudstack/service/NsxApiClient.java [hunk 9]
```diff
@@ -526,24 +557,37 @@ public void createStaticNatRule(String vpcName, String tier1GatewayName,
...
+            PolicyNatRule natRule = natService.get(tier1GatewayName, NatId.USER.name(), ruleName);
+            if (natRule != null && !natRule.getMarkedForDelete()) {
+                logger.debug(String.format("Deleting rule %s from Tier 1 Gateway %s", ruleName, tier1GatewayName));
```
Comment (DaanHoogland) [inline plugins/network-elements/nsx/src/main/java/org/apache/cloudstack/service/NsxApiClient.java:581]:
```suggestion
                logger.debug("Deleting rule {} from Tier 1 Gateway {}", ruleName, tier1GatewayName);
```

File: plugins/network-elements/nsx/src/main/java/org/apache/cloudstack/service/NsxApiClient.java [hunk 10]
```diff
@@ -577,9 +621,14 @@ public boolean doesPfRuleExist(String ruleName, String tier1GatewayName) {
         try {
             NatRules natService = (NatRules) nsxService.apply(NatRules.class);
             PolicyNatRule rule = natService.get(tier1GatewayName, NAT_ID, ruleName);
+            logger.debug(String.format("Rule %s from Tier 1 GW %s: %s", ruleName, tier1GatewayName,
+                    rule == null ? "null" : rule.getId() + " " + rule.getPath()));
```
Comment (DaanHoogland) [inline plugins/network-elements/nsx/src/main/java/org/apache/cloudstack/service/NsxApiClient.java:625]:
```suggestion
            logger.debug("Rule {} from Tier 1 GW {}: {}", ruleName, tier1GatewayName,
                    rule == null ? "null" : rule.getId() + " " + rule.getPath());
```

File: plugins/network-elements/nsx/src/main/java/org/apache/cloudstack/utils/NsxControllerUtils.java [hunk 11]
```diff
@@ -124,6 +124,9 @@ public static String getServerPoolName(String tier1GatewayName, long lbId) {
     }
 
     public static String getActiveMonitorProfileName(String lbServerPoolName, String port, String protocol) {
+        if (protocol.equalsIgnoreCase("udp")) {
+            protocol =  "ICMP";
+        }
```
Comment (DaanHoogland) [inline plugins/network-elements/nsx/src/main/java/org/apache/cloudstack/utils/NsxControllerUtils.java:129]: why is udp forbidden and why is it replaced by icmp? (next icmp + port is returned which is not realy a thing)

File: server/src/main/java/com/cloud/network/NetworkServiceImpl.java [hunk 12]
```diff
@@ -1752,7 +1752,7 @@ public Network createGuestNetwork(CreateNetworkCmd cmd) throws InsufficientCapac
 
     private void validateNetworkCreationSupported(long zoneId, String zoneName, GuestType guestType) {
         NsxProviderVO nsxProviderVO = nsxProviderDao.findByZoneId(zoneId);
-        if (Objects.nonNull(nsxProviderVO) && List.of(GuestType.L2, GuestType.Shared).contains(guestType)) {
+        if (Objects.nonNull(nsxProviderVO) && List.of(GuestType.L2).contains(guestType)) {
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/network/NetworkServiceImpl.java:1755]:
```suggestion
        if (Objects.nonNull(nsxProviderVO) && GuestType.L2.equals(guestType)) {
```

File: server/src/main/java/com/cloud/network/vpc/VpcManagerImpl.java [hunk 13]
```diff
@@ -3161,17 +3170,21 @@ protected List<IPAddressVO> listPublicIpsAssignedToVpc(final long accountId, fin
...
 
         if (sourceNatIp != null) {
             ipToReturn = PublicIp.createFromAddrAndVlan(sourceNatIp, _vlanDao.findById(sourceNatIp.getVlanId()));
         } else {
-            ipToReturn = _ipAddrMgr.assignDedicateIpAddress(owner, null, vpc.getId(), dcId, true);
+            if (!forNsx) {
+                ipToReturn = _ipAddrMgr.assignDedicateIpAddress(owner, null, vpc.getId(), dcId, true);
+            } else {
+                ipToReturn = _ipAddrMgr.assignPublicIpAddress(dcId, podId, owner, Vlan.VlanType.VirtualNetwork, null, null, false, true);
+            }
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/network/vpc/VpcManagerImpl.java:3187]: negative condition first is not optimal
```suggestion
            if (forNsx) {
                ipToReturn = _ipAddrMgr.assignPublicIpAddress(dcId, podId, owner, Vlan.VlanType.VirtualNetwork, null, null, false, true);
            } else {
                ipToReturn = _ipAddrMgr.assignDedicateIpAddress(owner, null, vpc.getId(), dcId, true);
            }
```

File: plugins/network-elements/nsx/src/main/java/org/apache/cloudstack/service/NsxElement.java [hunk 14]
```diff
@@ -527,45 +544,77 @@ public boolean applyStaticNats(Network config, List<? extends StaticNat> rules)
...
+                        .setPrivatePort(privatePort)
+                        .setPublicPort(publicPort)
+                        .setRuleId(rule.getId())
+                        .setProtocol(rule.getProtocol().toUpperCase(Locale.ROOT))
+                        .build();
+                FirewallRuleDetailVO ruleDetail = firewallRuleDetailsDao.findDetail(rule.getId(), ApiConstants.FOR_NSX);
+                if (Arrays.asList(FirewallRule.State.Add, FirewallRule.State.Active).contains(rule.getState())) {
+                    if ((ruleDetail == null && FirewallRule.State.Add == rule.getState()) || (ruleDetail != null && !ruleDetail.getValue().equalsIgnoreCase("true"))) {
+                        logger.debug(String.format("Creating port forwarding rule on NSX for VM %s to ports %s - %s",
+                                vm.getUuid(), rule.getDestinationPortStart(), rule.getDestinationPortEnd()));
```
Comment (DaanHoogland) [inline plugins/network-elements/nsx/src/main/java/org/apache/cloudstack/service/NsxElement.java:580]:
```suggestion
                        logger.debug("Creating port forwarding rule on NSX for VM {} to ports {} - {}",
                                vm.getUuid(), rule.getDestinationPortStart(), rule.getDestinationPortEnd());
```

File: plugins/network-elements/nsx/src/main/java/org/apache/cloudstack/service/NsxElement.java [hunk 15]
```diff
@@ -527,45 +544,77 @@ public boolean applyStaticNats(Network config, List<? extends StaticNat> rules)
...
+                        NsxAnswer answer = nsxService.createPortForwardRule(networkRule);
+                        boolean pfRuleResult = answer.getResult();
+                        if (pfRuleResult && !answer.isObjectExistent()) {
+                            logger.debug(String.format("Port forwarding rule %s created on NSX, adding detail on firewall rules details", rule.getId()));
```
Comment (DaanHoogland) [inline plugins/network-elements/nsx/src/main/java/org/apache/cloudstack/service/NsxElement.java:584]:
```suggestion
                            logger.debug("Port forwarding rule {} created on NSX, adding detail on firewall rules details", rule.getId());
```

File: plugins/network-elements/nsx/src/main/java/org/apache/cloudstack/service/NsxElement.java [hunk 16]
```diff
@@ -527,45 +544,77 @@ public boolean applyStaticNats(Network config, List<? extends StaticNat> rules)
...
+                            if (ruleDetail == null && FirewallRule.State.Add == rule.getState()) {
+                                logger.debug(String.format("Adding new firewall detail for rule %s", rule.getId()));
```
Comment (DaanHoogland) [inline plugins/network-elements/nsx/src/main/java/org/apache/cloudstack/service/NsxElement.java:586]:
```suggestion
                                logger.debug("Adding new firewall detail for rule {}", rule.getId());
```

File: plugins/network-elements/nsx/src/main/java/org/apache/cloudstack/service/NsxElement.java [hunk 17]
```diff
@@ -527,45 +544,77 @@ public boolean applyStaticNats(Network config, List<? extends StaticNat> rules)
...
+                                firewallRuleDetailsDao.addDetail(rule.getId(), ApiConstants.FOR_NSX, "true", false);
+                            } else {
+                                logger.debug(String.format("Updating firewall detail for rule %s", rule.getId()));
```
Comment (DaanHoogland) [inline plugins/network-elements/nsx/src/main/java/org/apache/cloudstack/service/NsxElement.java:589]:
```suggestion
                                logger.debug("Updating firewall detail for rule ", rule.getId());
```

File: plugins/network-elements/nsx/src/main/java/org/apache/cloudstack/service/NsxElement.java [hunk 18]
```diff
@@ -527,45 +544,77 @@ public boolean applyStaticNats(Network config, List<? extends StaticNat> rules)
...
+                                firewallRuleDetailsDao.update(ruleDetail.getId(), ruleDetail);
+                            }
+                        }
+                        result &= pfRuleResult;
+                    }
+                } else if (rule.getState() == FirewallRule.State.Revoke) {
+                    if (ruleDetail == null || (ruleDetail != null && ruleDetail.getValue().equalsIgnoreCase("true"))) {
+                        boolean pfRuleResult = nsxService.deletePortForwardRule(networkRule);
+                        if (pfRuleResult && ruleDetail != null) {
+                            logger.debug(String.format("Updating firewall rule detail %s for rule %s, set to false", ruleDetail.getId(), rule.getId()));
```
Comment (DaanHoogland) [inline plugins/network-elements/nsx/src/main/java/org/apache/cloudstack/service/NsxElement.java:600]:
```suggestion
                            logger.debug("Updating firewall rule detail {} for rule {}, set to false", ruleDetail.getId(), rule.getId());
```

File: plugins/network-elements/nsx/src/main/java/org/apache/cloudstack/utils/NsxControllerUtils.java (hunk 11 above)
Comment (nvazquez) [reply plugins/network-elements/nsx/src/main/java/org/apache/cloudstack/utils/NsxControllerUtils.java:129]: This was a workaround to an issue found while enabling maintenance on an NSX edge node, after disabling maintenance the UDP load balancer wouldn't come up again as the active monitor failed instantiating, so the workaround was creating ICMP active monitors for UDP load balancers

File: plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/actionworkers/KubernetesClusterResourceModifierActionWorker.java (hunk 4 above)
Comment (nvazquez) [reply plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/actionworkers/KubernetesClusterResourceModifierActionWorker.java]: Thanks, refactored

File: server/src/main/java/com/cloud/network/vpc/VpcManagerImpl.java (hunk 13 above)
Comment (nvazquez) [reply server/src/main/java/com/cloud/network/vpc/VpcManagerImpl.java:3187]: @DaanHoogland no problem I'll refactor it, but why isn't it optimal?

File: engine/components-api/src/main/java/com/cloud/network/vpc/VpcManager.java (hunk 3 above)
Comment (nvazquez) [reply engine/components-api/src/main/java/com/cloud/network/vpc/VpcManager.java]: @DaanHoogland thanks I've refactored it to remove the boolean from the service interface

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by rohityadavcloud** [comment]
@blueorangutan test matrix

**Comment by rohityadavcloud** [comment]
@Pearl1594 @nvazquez do we need manual testing for this, can you review and address further on this PR?

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test matrix

**Comment by alexandremattioli** [comment]
LGTM

**Review by alexandremattioli — APPROVED** [review]

**Review by alexandremattioli — APPROVED** [review]
LGTM

**Comment by nvazquez** [comment]
Hi @rohityadavcloud this PR is ready (waiting for test results)

**Comment by rohityadavcloud** [comment]
Just to check build:
@blueorangutan package

**Comment by DaanHoogland** [comment]
@Pearl1594 , how critical are these fixes ?

**Comment by rohityadavcloud** [comment]
@blueorangutan test ol8 vmware-70u3
